crates/burn-import/src/burn/node/gather.rs

```diff
@@ -27,33 +27,79 @@ impl<PS: PrecisionSettings> NodeCodegen<PS> for GatherNode {
         node_position: usize,
     ) -> proc_macro2::TokenStream {
         let dim = self.dim.to_tokens();
+        let input_rank = match &self.input {
+            Type::Tensor(in_tensor) => in_tensor.dim,
+            Type::Shape(_) => 1,
+            _ => panic!("Gather needs Tensor or Shape input, got {:?}!", self.input),
+        };
+        let index_rank = match &self.index {
+            Type::Tensor(idx_tensor) => idx_tensor.dim,
+            Type::Scalar(_) => 0,
+            _ => panic!("Gather needs Scalar or Tensor index, got {:?}!", self.index),
+        };
+        let output_rank = index_rank + input_rank - 1;
+
         let input = match &self.input {
             Type::Tensor(in_tensor) => scope.tensor_use_owned(in_tensor, node_position),
             Type::Shape(in_shape) => {
                 let in_shape_name = &in_shape.name;
                 // To copy just the values from the shape value without moving it
                 // (which could lead to ownership problems if the same Shape is used multiple times)
                 // borrow the array as a slice and use that to create the Tensor:
-                quote! { Tensor::from_data(&#in_shape_name as &[_], &*self.device) }
+                quote! { Tensor::<B, 1, Int>::from_data(&#in_shape_name as &[_], &*self.device) }
             }
             _ => panic!("Gather needs Scalar or Shape input, got {:?}!", self.input),
         };
+
         let output = &self.output.name;
 
+        let gather = match &self.index {
```

Comment:
Instead of having two `match` statements on `self.index` one after the other, we should combine the logic for the match arms.